Refactor routes

benbovy opened the pull request:
First step towards addressing #25.

This moves all path operation functions out of `RestAccessor` and creates instead `fastapi.APIRouter` instances in a new `routers` sub-package. Each module in `routers` contains a `APIRouter` instance dedicated to a specific part of the API. 

Each function operates on the served dataset by overriding the `get_dataset` dependency for `RestAccessor.app`.

TODO:

- [x] move zarr-specific path operation functions after #21
- ~~maybe refactor tests (if directly testing `APIRouter` instances is possible and a good idea)~~

benbovy commented:
@jhamman / @lsetiawan could you have a look? This is ready for review.

It is mainly a reorganization of existing blocks of code, so there is not much additional logic implemented here. Also, there is no breaking change in the REST API.

I have wrapped the REST cache in its own Dataset accessor (`Dataset._rest_cache`). It is currently for internal use, but it might be good to expose it publicly, e.g., if we later want to allow custom API endpoints reusing the cache. For convenience, it could still be accessed from the main accessor `Dataset.rest`, although I'm not sure if this is really needed?

I've done the same thing for the zarr layer with a new `Dataset._rest_zarr` internal accessor. Not sure if there is anything there that is intended to be exposed publicly. `Dataset._rest_zarr.zmetadata` or other things, maybe?

From here, it will be straightforward to implement custom api endpoints and add support for serving multiple datasets with a single FastAPI application (in follow-up PRs).

lsetiawan commented:
@benbovy This is really great! Extending the xarray dataset stuff is new to me. Good to see how it's used. I think separating internal accessors and routes are awesome! :smile:

benbovy commented:
> I have wrapped the REST cache in its own Dataset accessor (Dataset._rest_cache)

Alternatively, we could expose the cache using a FastAPI dependency (overridden when creating the app), just like for accessing the dataset. This might be a better option if we want more flexible control, e.g., use a single cache for serving a collection of datasets.

lsetiawan commented:
> use a single cache for serving a collection of datasets.
Is there a way to use the dataset id in a collection as part of the cache_key maybe? This way the cache is centralized.

benbovy commented:
> Is there a way to use the dataset id in a collection as part of the cache_key maybe?

Yes I guess it would be pretty easy to do that. For multiple published datasets we might need a thin wrapper around `cachey.Cache` in order to prepend the dataset id to the cache_key. Or we could simply check if the `dataset_id` parameter exists and/or is not None to support both single vs multiple published datasets (although I don't know if it easy to do that with FastAPI).

benbovy commented:
Now I wonder if we could get rid of the `Dataset._rest_zarr` accessor. The only benefit I see is that `zmetadata` is cached, but maybe we could just reuse the (cachey) cache for that? Or is there any motivation to expose `zmetadata` publicly in xpublish?

benbovy commented:
> Now I wonder if we could get rid of the Dataset._rest_zarr accessor

I ended up doing it. It now follows a bit more FastAPI idioms and the Dataset namespace is less polluted. Nothing really important though...

benbovy commented:
The last commit was needed in order to move "/info" from the zarr router into the base router, which makes more sense.

The refactoring gets pretty big, but I think I'll stop here as I'm pretty happy with it now :-). The code has been rearranged into many small modules, which may seem a bit too much at this stage, but I think this structure will really help to keep the code tidy as the project grows.

jhamman commented:
Just taking a look at this now. Will leave some detailed comments but this looks great.

benbovy commented:
Thanks for your review @jhamman.

I was thinking of addressing your comments in a follow-up PR (custom routers) to avoid adding too much things here.

> Updates to the docs (probably the tutorial) showing how you would add a custom router

We'll probably need to update xpublish's description as well, as the Zarr compatible REST API becomes one of the possible features enabled by xpublish, among others.

> What would be the recommended way of setting a custom router?

One way to customize routes is via a optional `routers` parameter in `RestAccessor.__call__`:

```python
my_dataset.rest(routers=None).serve()
```

`routers` would accept a list of `APIRouters` objects (or `None` for default routes). Additionally, we might want to allow passing some (or any) keyword arguments to `app.include_router`, e.g., using a `Tuple[APIRouter, Dict]`:

```python
foo = APIRouter()
bar = APIRouter()

my_dataset.rest(
    routers=[foo, (bar, {'prefix': 'bar'})]
).serve()
```

This would be useful for setting custom tags, prefixes or dependencies. For example, one may want to expose the Zarr compatible REST API with a `zarr` prefix and/or only to authenticated API users.

> Do we understand how namespace conflicts are handled here? 

It looks like FastAPI doesn't raise any error in case of a conflict but instead uses the last defined/added function. It would be better if we explicitly check for conflicts here in my opinion.

benbovy commented on tests/utils.py:
```diff
@@ -21,6 +22,20 @@ def __getitem__(self, key):
         return response.content
 
 
+class DummyCache(cachey.Cache):
```
I was used but not anymore. Removed.

jhamman commented:
Okay, I like the idea of pushing the majority of my comments to follow up PRs. I'm going to merge now 🎉
